### ISSUE TITLE
No diagnostics in neovim 0.11 with native lsp

### ISSUE THREAD

**Issue by maxstollmayer** [astral-sh/ty#2616]
### Question

I want to tryout ty in my neovim 0.11.5 setup using the native lsp, but cannot seem to get it to report any issues. Using `ty check` works just fine and also my config for ruff works in neovim.

The minimal setup from "Editor integration" does not work for me, since it misses the `cmd` setting. Below is the minimal `init.lua` that I could get to work for ruff, but it does not work for ty.
```lua
vim.lsp.config("ty", {
    cmd = { "ty", "server" }
})
vim.lsp.enable("ty")

vim.lsp.config("ruff", {
    cmd = { "ruff", "server" }
})
vim.lsp.enable("ruff")
```

In `main.py` I induced both a syntax and type error:
```python
def main() -> str:
    print("hello"
```

Using `ty check main.py` these errors get reported correctly and ruff also reports the syntax error in neovim, but ty does not. Here is the trace log from that test: [ty.log](https://github.com/user-attachments/files/24839029/ty.log)
I also tried to use `uvx ty server` as the command and also `uv run ty server` with ty installed in the project but the result is the same.

Am I doing something wrong in this setup? Thanks for any help on this as I would love to try out ty.

### Version

ty 0.0.13 (fc1478bd9 2026-01-21)

**Comment by dhruvmanila** [comment]
Thank you for the reporting this issue!

Can you provide some additional information like the current working directory from where you're opening Neovim and the project structure?

I don't think you're doing anything wrong here so it could be something related to the way Neovim is being opened and / or the project structure.

**Comment by maxstollmayer** [comment]
Thanks for taking a look!

For this minimal example it was basically just what uv creates with `uv init test`:

```
test/
│   .python-version
│   main.py
│   pyproject.toml
│   README.md
│   uv.lock
├───.venv/
└───__pycache__/
```

I open this whole folder as the working directory in neovim using `nvim .` in Windows Terminal from my Windows 11 PC. Since ruff works just fine I am not sure that the working directory would be wrong, unless ty handles that very differently.

**Comment by loganthomas** [comment]
Is it possible that `ty` needs the file to exist on disk to analyze it? I've found that i won't see any errors  until i `:wq` and then reopen my file.

Upon file creation (`nvim test.py`)
<img width="1105" height="531" alt="Image" src="https://github.com/user-attachments/assets/34e0aa2c-5aeb-473a-b0b2-54420828b8d0" />

Post `:wq` followed by `nvim test.py`
<img width="1096" height="525" alt="Image" src="https://github.com/user-attachments/assets/27f21207-6df0-4221-bae4-0c811df7e11e" />

Environment:
```
  ty: 0.0.14 (16597f5e2 2026-01-26)
  Neovim: v0.11.6
  macOS: 14.6.1
  MacBook Pro, Intel Core i7-8850H @ 2.60GHz, 32 GB RAM
```

**Comment by maxstollmayer** [comment]
Thanks, it may be so, but this is not the (sole) issue here. During my testing there have been several files that were created and have existed, during many restarts of neovim and even reboots in various states of erroneous to error-free python.

**Comment by BurntSushi** [comment]
All righty, so I tried to reproduce this in a barebones config, and I can't seem to. Here's what I did.

Inside a scratch directory, I ran:

```
uv init test
cd test
```

Then I made `main.py` match yours:

```
$ cat > main.py <<EOF
heredoc> def main() -> str:
    print("hello"
heredoc> EOF
$ cat main.py
def main() -> str:
    print("hello"
```

Next, I [installed `nvim-lspconfig`](https://github.com/neovim/nvim-lspconfig?tab=readme-ov-file#install):

```
git clone https://github.com/neovim/nvim-lspconfig ~/.config/nvim/pack/nvim/start/nvim-lspconfig
```

And now I set up a barebones neovim config:

```
$ cat > init.lua <<EOF
vim.lsp.config('ty', {})
vim.lsp.enable('ty')
EOF
```

And finally, open neovim using that barebones config:

```
nvim -u init.lua main.py
```

And this is what I get:

https://github.com/user-attachments/assets/90550902-2ae9-4b65-b72e-e96a6cc1a678

I used `<Control-W> d` to show the diagnostic floating window.

Here's my nvim version:

```
$ nvim --version
NVIM v0.11.6
Build type: RelWithDebInfo
LuaJIT 2.1.1767980792
Run "nvim -V1 -v" for more info
```

> The minimal setup from "Editor integration" does not work for me, since it misses the `cmd` setting.

Hmmm, [the nvim-lspconfig does indeed have a `cmd` setting](https://github.com/neovim/nvim-lspconfig/blob/3f58aeca0c6ece8a9fb8782ea3fcb6024f285be3/lsp/ty.lua#L11), and my minimal config above (which works for me) does not need to set it. Maybe there is something else going wrong here.

**Comment by ypomortsev** [comment]
Using @BurntSushi's barebones config, I'm getting diagnostics for existing files (`main.py`) but not for new files, even after saving the file and reloading it:

https://github.com/user-attachments/assets/e80ad59e-706a-4b8a-b0f2-e89485ac7862

I'm also using the same nvim version:

```
NVIM v0.11.6
Build type: RelWithDebInfo
LuaJIT 2.1.1767980792
Run "nvim -V1 -v" for more info
```

**Comment by BurntSushi** [comment]
@ypomortsev Ah I can indeed reproduce that! Thank you. I'll take a look at this soon.

**Comment by BurntSushi** [comment]
I believe https://github.com/astral-sh/ruff/pull/23095 will fix this problem. But you'll need to enable `workspace/didChangeWatchedFiles` _explicitly_ if you're on Linux:

```lua
local capabilities = vim.lsp.protocol.make_client_capabilities()
capabilities.workspace.didChangeWatchedFiles.dynamicRegistration = true

vim.lsp.config('ty', {
  capabilities = capabilities,
})
vim.lsp.enable('ty')
```

**Comment by BurntSushi** [comment]
One other tip that I use for other LSPs too is to add a function to your neovim config for restarting the server:

```lua
function RestartLSP()
    print('Restarting all attached LSP clients and reloading buffer...')
    vim.lsp.stop_client(vim.lsp.get_clients())
    vim.cmd('write')
    vim.cmd('edit')
end
```

Then call it with `:lua RestartLSP()`. That in particular would help work around issues like this when you come across them (without restarting your entire editor).

**Comment by maxstollmayer** [comment]
Thanks for looking into this @BurntSushi. Since I am on neovim 0.11+ I want to use the new [native LSP setup](https://neovim.io/doc/user/lsp.html#vim.lsp.config()) without the `nvim-lspconfig` plugin you used. The config mentioned in your [documentation](https://docs.astral.sh/ty/editors/#neovim) does not work for me because the new neovim LSP setup requires a different config as mentioned in my post above. Since ruff and all other LSPs that I use work just fine with this new LSP setup I was hoping that ty would too. But I am not sure if I need to configure it differently for the new LSP setup or if ty needs to do something differently in this setup.

**Comment by BurntSushi** [comment]
My config above is using the new neovim 0.11+ stuff. Note that neovim 0.11 does not obviate nvim-lspconfig. You can see that here: https://github.com/neovim/nvim-lspconfig?tab=readme-ov-file#important-%EF%B8%8F

**Comment by BurntSushi** [comment]
> The config mentioned in your [documentation](https://docs.astral.sh/ty/editors/#neovim) does not work for me because the new neovim LSP setup requires a different config as mentioned in my post above.

That link has the old, but right below it mentions the neovim 0.11+ way to go about it.

**Comment by maxstollmayer** [comment]
That is true. But the config that nvim-lspconfig provides is the minimal config that I have manually setup anyway: [https://github.com/neovim/nvim-lspconfig/blob/master/lsp/ty.lua](https://github.com/neovim/nvim-lspconfig/blob/master/lsp/ty.lua) So I will not install nvim-lspconfig since it should work without, but I still do not get any diagnostics after upgrading everything to the latest version. I am happy to hear more suggestions though or if you need any tests I could do to help, let me know.

**Comment by maxstollmayer** [comment]
> > The config mentioned in your [documentation](https://docs.astral.sh/ty/editors/#neovim) does not work for me because the new neovim LSP setup requires a different config as mentioned in my post above.
> 
> That link has the old, but right below it mentions the neovim 0.11+ way to go about it.

Yes that is what I tried to mention, but the way it reads to me is that I do not need nvim-lspconfig. But then that given config does not work at all since it assumes the `cmd` parameter is given via nvim-lspconfig.

**Comment by BurntSushi** [comment]
Yes, for the specific case of newly created buffers in the same editor session, this PR should fix that: https://github.com/astral-sh/ruff/pull/23095 --- It isn't merged yet.

Diagnostics should otherwise be working. Are they not for you? Basically, if you're still not getting diagnostics _after_ an editor restart for files already on disk, then that represents a problem I haven't been able to reproduce yet. If so, can you give more detailed instructions for how to do that?

Can you also please try with `nvim-lspconfig` as a debugging step? `nvim-lspconfig` has a lot of stuff in it, and I'm actually not totally clear how much work it's actually doing. It's probably very low priority for us to make ty work in neovim without `nvim-lspconfig`.

**Comment by maxstollmayer** [comment]
That problem came up during the discussion of my initial issue of no diagnostics at all. If you need further info than my original question from above including the attached log file, please let me know.

I will try with nvim-lspconfig in the next few days and report back.

**Comment by maxstollmayer** [comment]
@BurntSushi can you please reopen this issue. The fix you closed this with is from a tangential problem that was raised in the comments and does not address this ticket itself, please see the original post from the top. The question still remains open and even using nvim-lspconfig I get no diagnostics from ty, whereas ruff does provide them. I have attached the [trace log](https://github.com/user-attachments/files/25215541/ty.log) from ty using the minimal nvim-lspconfig setup `vim.lsp.enable("ty")`. I am happy to receive any hints.

**Comment by BurntSushi** [comment]
Yeah I can't tell what's going wrong for you. Without something that I can reproduce, I'm not quite sure how to help. Nothing really stands out for me in the log. ty is clearly picking up your `main.py`.

And I can also verify that things work in this simple case even without `nvim-lspconfig`. In a directory, I have an `init.lua`:

```lua
$ cat init.lua
vim.lsp.config('ty', {
  cmd = {'ty', 'server'},
})
vim.lsp.enable('ty')
```

A `pyproject.toml`:

```toml
[project]
name = "test"
version = "0.1.0"
requires-python = ">=3.13"
dependencies = []
```

A `main.py`:

```python
def main() -> str:
    print("hello"
```

Then I do `nvim -u init.lua main.py` and here's what I see:

https://github.com/user-attachments/assets/82a20acf-892a-438f-b8d8-471c7d02c4b5

This is on latest main at commit 6d797e236f49c1921c6a8e0ef555e44ca5ce3db3.

I noticed from the logs that you're on Windows. I'm on Linux. I think that's the only other variable here that I'm not testing. It's not easy for me to test on Windows.

**Comment by maxstollmayer** [comment]
OK, thanks for trying anyway. Maybe it is Windows specific then.

**Comment by dhruvmanila** [comment]
@maxstollmayer Can you provide the Neovim LSP's debug logs i.e., the data that flows between the client and the server? You can enable it via:

```lua
vim.lsp.log.set_level(vim.log.levels.DEBUG)
```

You can get the file path where these logs are going using `vim.lsp.log.get_filename()`. Make sure to clear the content of this file before going through the reproduction steps to only record the logs that are part of this issue (if possible).

You'll see something like this (very detailed logs):

<details><summary>Details</summary>
<p>

```
[START][2026-02-24 11:45:40] LSP logging initiated
2026-02-24 11:45:40 [INFO] Starting RPC client {
  cmd = { "/Users/dhruv/work/astral/ruff/target/debug/ty", "server" },
  extra = {}
}
2026-02-24 11:45:40 [DEBUG] rpc.send {
  id = 1,
  jsonrpc = "2.0",
  method = "initialize",
  params = {
    capabilities = {
      general = {
        positionEncodings = { "utf-8", "utf-16", "utf-32" }
      },
      textDocument = {
        callHierarchy = {
          dynamicRegistration = false
        },
        codeAction = {
          codeActionLiteralSupport = {
            codeActionKind = {
              valueSet = { "", "quickfix", "refactor", "refactor.extract", "refactor.inline", "refactor.rewrite", "source", "source.organizeImports" }
            }
          },
          dataSupport = true,
          disabledSupport = true,
          dynamicRegistration = true,
          honorsChangeAnnotations = true,
          isPreferredSupport = true,
          resolveSupport = {
            properties = { "edit", "command" }
          }
        },
        codeLens = {
          dynamicRegistration = false,
          resolveSupport = {
            properties = { "command" }
          }
        },
        colorProvider = {
          dynamicRegistration = true
        },
        completion = {
          completionItem = {
            commitCharactersSupport = false,
            deprecatedSupport = true,
            documentationFormat = { "markdown", "plaintext" },
            insertReplaceSupport = true,
            insertTextModeSupport = {
              valueSet = { 1 }
            },
            labelDetailsSupport = true,
            preselectSupport = false,
            resolveSupport = {
              properties = { "documentation", "detail", "additionalTextEdits", "command", "data" }
            },
            snippetSupport = true,
            tagSupport = {
              valueSet = { 1 }
            }
          },
          completionItemKind = {
            valueSet = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25 }
          },
          completionList = {
            itemDefaults = { "commitCharacters", "editRange", "insertTextFormat", "insertTextMode", "data" }
          },
          contextSupport = true,
          dynamicRegistration = false,
          insertTextMode = 1
        },
        declaration = {
          linkSupport = true
        },
        definition = {
          dynamicRegistration = true,
          linkSupport = true
        },
        diagnostic = {
          dataSupport = true,
          dynamicRegistration = true,
          relatedDocumentSupport = true,
          relatedInformation = true,
          tagSupport = {
            valueSet = { 1, 2 }
          }
        },
        documentHighlight = {
          dynamicRegistration = false
        },
        documentSymbol = {
          dynamicRegistration = false,
          hierarchicalDocumentSymbolSupport = true,
          symbolKind = {
            valueSet = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26 }
          },
          tagSupport = {
            valueSet = { 1 }
          }
        },
        foldingRange = {
          dynamicRegistration = false,
          foldingRange = {
            collapsedText = true
          },
          foldingRangeKind = {
            valueSet = { "comment", "imports", "region" }
          },
          lineFoldingOnly = true
        },
        formatting = {
          dynamicRegistration = true
        },
        hover = {
          contentFormat = { "markdown", "plaintext" },
          dynamicRegistration = true
        },
        implementation = {
          linkSupport = true
        },
        inlayHint = {
          dynamicRegistration = true,
          resolveSupport = {
            properties = { "textEdits", "tooltip", "location", "command" }
          }
        },
        inlineCompletion = {
          dynamicRegistration = false
        },
        linkedEditingRange = {
          dynamicRegistration = false
        },
        onTypeFormatting = {
          dynamicRegistration = false
        },
        publishDiagnostics = {
          dataSupport = true,
          relatedInformation = true,
          tagSupport = {
            valueSet = { 1, 2 }
          }
        },
        rangeFormatting = {
          dynamicRegistration = true,
          rangesSupport = true
        },
        references = {
          dynamicRegistration = false
        },
        rename = {
          dynamicRegistration = true,
          honorsChangeAnnotations = true,
          prepareSupport = true
        },
        selectionRange = {
          dynamicRegistration = false
        },
        semanticTokens = {
          augmentsSyntaxTokens = true,
          dynamicRegistration = false,
          formats = { "relative" },
          multilineTokenSupport = true,
          overlappingTokenSupport = true,
          requests = {
            full = {
              delta = true
            },
            range = true
          },
          serverCancelSupport = false,
          tokenModifiers = { "declaration", "definition", "readonly", "static", "deprecated", "abstract", "async", "modification", "documentation", "defaultLibrary" },
          tokenTypes = { "namespace", "type", "class", "enum", "interface", "struct", "typeParameter", "parameter", "variable", "property", "enumMember", "event", "function", "method", "macro", "keyword", "modifier", "comment", "string", "number", "regexp", "operator", "decorator" }
        },
        signatureHelp = {
          dynamicRegistration = false,
          signatureInformation = {
            activeParameterSupport = true,
            documentationFormat = { "markdown", "plaintext" },
            noActiveParameterSupport = true,
            parameterInformation = {
              labelOffsetSupport = true
            }
          }
        },
        synchronization = {
          didSave = true,
          dynamicRegistration = false,
          willSave = true,
          willSaveWaitUntil = true
        },
        typeDefinition = {
          linkSupport = true
        }
      },
      window = {
        showDocument = {
          support = true
        },
        showMessage = {
          messageActionItem = {
            additionalPropertiesSupport = true
          }
        },
        workDoneProgress = true
      },
      workspace = {
        applyEdit = true,
        codeLens = {
          refreshSupport = true
        },
        configuration = true,
        diagnostics = {
          refreshSupport = true
        },
        didChangeConfiguration = {
          dynamicRegistration = false
        },
        didChangeWatchedFiles = {
          dynamicRegistration = false,
          relativePatternSupport = true
        },
        inlayHint = {
          refreshSupport = true
        },
        semanticTokens = {
          refreshSupport = true
        },
        symbol = {
          dynamicRegistration = false,
          symbolKind = {
            valueSet = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26 }
          }
        },
        workspaceEdit = {
          changeAnnotationSupport = {
            groupsOnLabel = true
          },
          normalizesLineEndings = true,
          resourceOperations = { "rename", "create", "delete" }
        },
        workspaceFolders = true
      }
    },
    clientInfo = {
      name = "Neovim",
      version = "0.12.0-dev+g886efcb853"
    },
    initializationOptions = {
      logFile = "/Users/dhruv/.local/state/nvim/lsp.ty.log",
      logLevel = "debug"
    },
    processId = 50575,
    rootPath = "/Users/dhruv/playground/ty-server",
    rootUri = "file:///Users/dhruv/playground/ty-server",
    trace = "off",
    workDoneToken = "1",
    workspaceFolders = { {
        name = "/Users/dhruv/playground/ty-server",
        uri = "file:///Users/dhruv/playground/ty-server"
      } }
  }
}
2026-02-24 11:45:40 [DEBUG] rpc.receive {
  id = 1,
  jsonrpc = "2.0",
  result = {
    capabilities = {
      codeActionProvider = {
        codeActionKinds = { "quickfix" }
      },
      completionProvider = {
        triggerCharacters = { "." }
      },
      declarationProvider = true,
      definitionProvider = true,
      documentHighlightProvider = true,
      documentSymbolProvider = true,
      executeCommandProvider = {
        commands = { "ty.printDebugInformation" },
        workDoneProgress = false
      },
      foldingRangeProvider = true,
      hoverProvider = true,
      inlayHintProvider = vim.empty_dict(),
      notebookDocumentSync = {
        notebookSelector = { {
            cells = { {
                language = "python"
              } }
          } },
        save = false
      },
      positionEncoding = "utf-8",
      referencesProvider = true,
      renameProvider = {
        prepareProvider = true
      },
      selectionRangeProvider = true,
      semanticTokensProvider = {
        full = true,
        legend = {
          tokenModifiers = { "definition", "readonly", "async", "documentation" },
          tokenTypes = { "namespace", "class", "parameter", "selfParameter", "clsParameter", "variable", "property", "function", "method", "keyword", "string", "number", "decorator", "builtinConstant", "typeParameter" }
        },
        range = true
      },
      signatureHelpProvider = {
        retriggerCharacters = { ")" },
        triggerCharacters = { "(", "," }
      },
      textDocumentSync = {
        change = 2,
        openClose = true
      },
      typeDefinitionProvider = true,
      workspace = {
        workspaceFolders = {
          changeNotifications = true,
          supported = true
        }
      },
      workspaceSymbolProvider = true
    },
    serverInfo = {
      name = "ty",
      version = "ruff/0.15.2+16 (bfcddb6ad 2026-02-20)"
    }
  }
}
2026-02-24 11:45:40 [DEBUG] rpc.send {
  jsonrpc = "2.0",
  method = "initialized",
  params = vim.empty_dict()
}
2026-02-24 11:45:40 [DEBUG] rpc.send {
  jsonrpc = "2.0",
  method = "textDocument/didOpen",
  params = {
    textDocument = {
      languageId = "python",
      text = "def foo(x: tuple[int, ...]):\n    (a, *b) = x\n",
      uri = "file:///Users/dhruv/playground/ty-server/src/ty_server/inlay_hints.py",
      version = 0
    }
  }
}
2026-02-24 11:45:40 [INFO] LSP[ty] server_capabilities {
  server_capabilities = {
    codeActionProvider = {
      codeActionKinds = { "quickfix" }
    },
    completionProvider = {
      triggerCharacters = { "." }
    },
    declarationProvider = true,
    definitionProvider = true,
    documentHighlightProvider = true,
    documentSymbolProvider = true,
    executeCommandProvider = {
      commands = { "ty.printDebugInformation" },
      workDoneProgress = false
    },
    foldingRangeProvider = true,
    hoverProvider = true,
    inlayHintProvider = vim.empty_dict(),
    notebookDocumentSync = {
      notebookSelector = { {
          cells = { {
              language = "python"
            } }
        } },
      save = false
    },
    positionEncoding = "utf-8",
    referencesProvider = true,
    renameProvider = {
      prepareProvider = true
    },
    selectionRangeProvider = true,
    semanticTokensProvider = {
      full = true,
      legend = {
        tokenModifiers = { "definition", "readonly", "async", "documentation" },
        tokenTypes = { "namespace", "class", "parameter", "selfParameter", "clsParameter", "variable", "property", "function", "method", "keyword", "string", "number", "decorator", "builtinConstant", "typeParameter" }
      },
      range = true
    },
    signatureHelpProvider = {
      retriggerCharacters = { ")" },
      triggerCharacters = { "(", "," }
    },
    textDocumentSync = {
      change = 2,
      openClose = true
    },
    typeDefinitionProvider = true,
    workspace = {
      workspaceFolders = {
        changeNotifications = true,
        supported = true
      }
    },
    workspaceSymbolProvider = true
  }
}
2026-02-24 11:45:40 [DEBUG] rpc.receive {
  id = 0,
  jsonrpc = "2.0",
  method = "workspace/configuration",
  params = {
    items = { {
        scopeUri = "file:///Users/dhruv/playground/ty-server",
        section = "ty"
      } }
  }
}
2026-02-24 11:45:40 [DEBUG] LSP[ty] client.request 1 textDocument/semanticTokens/range {
  range = {
    ["end"] = {
      character = 0,
      line = 2
    },
    start = {
      character = 0,
      line = 0
    }
  },
  textDocument = {
    uri = "file:///Users/dhruv/playground/ty-server/src/ty_server/inlay_hints.py"
  }
} function: 0x010364ccf8 2
2026-02-24 11:45:40 [DEBUG] rpc.send {
  id = 2,
  jsonrpc = "2.0",
  method = "textDocument/semanticTokens/range",
  params = {
    range = {
      ["end"] = {
        character = 0,
        line = 2
      },
      start = {
        character = 0,
        line = 0
      }
    },
    textDocument = {
      uri = "file:///Users/dhruv/playground/ty-server/src/ty_server/inlay_hints.py"
    }
  }
}
2026-02-24 11:45:40 [DEBUG] LSP[ty] client.request 1 textDocument/semanticTokens/full {
  textDocument = {
    uri = "file:///Users/dhruv/playground/ty-server/src/ty_server/inlay_hints.py"
  }
} function: 0x010364f6c0 2
2026-02-24 11:45:40 [DEBUG] rpc.send {
  id = 3,
  jsonrpc = "2.0",
  method = "textDocument/semanticTokens/full",
  params = {
    textDocument = {
      uri = "file:///Users/dhruv/playground/ty-server/src/ty_server/inlay_hints.py"
    }
  }
}
2026-02-24 11:45:40 [DEBUG] server_request: callback result {
  result = { vim.NIL },
  status = true
}
2026-02-24 11:45:40 [DEBUG] rpc.send {
  id = 0,
  jsonrpc = "2.0",
  result = { vim.NIL }
}
2026-02-24 11:45:40 [DEBUG] rpc.receive {
  id = 1,
  jsonrpc = "2.0",
  method = "client/registerCapability",
  params = {
    registrations = { {
        id = "ty/textDocument/diagnostic",
        method = "textDocument/diagnostic",
        registerOptions = {
          documentSelector = vim.NIL,
          identifier = "ty",
          interFileDependencies = true,
          workDoneProgress = false,
          workspaceDiagnostics = false
        }
      } }
  }
}
2026-02-24 11:45:40 [DEBUG] LSP[ty] client.request 1 textDocument/diagnostic {
  textDocument = {
    uri = "file:///Users/dhruv/playground/ty-server/src/ty_server/inlay_hints.py"
  }
} function: 0x0102fd92f0 2
2026-02-24 11:45:40 [DEBUG] rpc.send {
  id = 4,
  jsonrpc = "2.0",
  method = "textDocument/diagnostic",
  params = {
    textDocument = {
      uri = "file:///Users/dhruv/playground/ty-server/src/ty_server/inlay_hints.py"
    }
  }
}
2026-02-24 11:45:40 [DEBUG] server_request: callback result {
  result = vim.NIL,
  status = true
}
2026-02-24 11:45:40 [DEBUG] rpc.send {
  id = 1,
  jsonrpc = "2.0",
  result = vim.NIL
}
2026-02-24 11:45:40 [DEBUG] rpc.receive {
  id = 2,
  jsonrpc = "2.0",
  result = {
    data = { 0, 4, 3, 7, 1, 0, 4, 1, 2, 1, 0, 3, 5, 1, 0, 0, 6, 3, 1, 0, 1, 5, 1, 5, 1, 0, 4, 1, 5, 1, 0, 5, 1, 2, 0 }
  }
}
2026-02-24 11:45:40 [DEBUG] rpc.receive {
  id = 3,
  jsonrpc = "2.0",
  result = {
    data = { 0, 4, 3, 7, 1, 0, 4, 1, 2, 1, 0, 3, 5, 1, 0, 0, 6, 3, 1, 0, 1, 5, 1, 5, 1, 0, 4, 1, 5, 1, 0, 5, 1, 2, 0 }
  }
}
2026-02-24 11:45:40 [DEBUG] rpc.receive {
  id = 4,
  jsonrpc = "2.0",
  result = {
    items = {},
    kind = "full"
  }
}
2026-02-24 11:45:40 [DEBUG] LSP[ty] client.request 1 textDocument/documentHighlight {
  position = {
    character = 0,
    line = 1
  },
  textDocument = {
    uri = "file:///Users/dhruv/playground/ty-server/src/ty_server/inlay_hints.py"
  }
} function: 0x0102fd8c08 2
2026-02-24 11:45:40 [DEBUG] rpc.send {
  id = 5,
  jsonrpc = "2.0",
  method = "textDocument/documentHighlight",
  params = {
    position = {
      character = 0,
      line = 1
    },
    textDocument = {
      uri = "file:///Users/dhruv/playground/ty-server/src/ty_server/inlay_hints.py"
    }
  }
}
2026-02-24 11:45:40 [DEBUG] rpc.receive {
  id = 5,
  jsonrpc = "2.0",
  result = vim.NIL
}
```

</p>
</details>
